.github/scripts/lint/lint.py

```diff
@@ -0,0 +1,174 @@
+#!/usr/bin/python3
+
+"""
+Custom linter for chisel slice definition files.
+
+Use in addition with yamllint (https://yamllint.readthedocs.io), since this
+script only covers slice definition file specifc rules.
```

Comment:
what are the specific rules that yamllint cannot cover and that this custom script does? I'd really like to avoid this complexity and just use yamllint

Reply:
This is to check the if the listed items in `slices`, `esstential` and `contents` are sorted alphabetically. I believe yamllint doesn't provide this already.

Reply:
You can do that with `yamllint`:

slices
```bash
yq '.slices | map_values("")' <SDF>.yaml | yamllint -d "{rules: {key-ordering: {}}}" -
```

essentials and contents
```bash
# can probably be improved with some more thinking
for slice in `yq '.slices | keys | .[]' <sdf>.yaml`
do
  echo "checking if essentials and contents are ordered in $slice"
  slice=$slice yq  '.slices | to_entries | .[] | select(.key == env(slice)) | .value | .essential[]' < <sdf>.yaml | sort -c
  slice=$slice yq  '.slices | to_entries | .[] | select(.key == env(slice)) | .value | .contents' < <sdf>.yaml | sort -c
done
```

Reply:
Nice yaml manipulation skills! Thanks!

So I wrote the following script to reflect your suggestions:

<details>
<summary> lint.sh </summary>

```sh
#!/bin/bash

# Custom linter for chisel slice definition files.
#
# Use in addition with yamllint (https://yamllint.readthedocs.io), since this
# script only covers slice definition file specifc rules.

set -eu

export LC_COLLATE=C

EXIT_CODE=0

err() {
	echo "error:" "$@" >&2
	EXIT_CODE=1
}

warn() {
	echo "warning:" "$@" >&2
}

check_slices_sorted() {
	local filename="$1"

	yq '.slices | keys | .[]' "$filename" | sort -C || \
		warn "$filename: slice names are not sorted"
}

check_essential_sorted() {
	local filename="$1"

	for slice in $(yq '.slices | keys | .[]' "$filename"); do
		key="$slice" yq \
			'.slices | with_entries(select(.key == env(key))) | .[].essential[]' \
			"$filename" | sort -C || \
			err "$filename: $slice: \"essential\" entries are not sorted"
	done
}

check_contents_sorted() {
	local filename="$1"

	for slice in $(yq '.slices | keys | .[]' "$filename"); do
		key="$slice" yq \
			'.slices | with_entries(select(.key == env(key))) | .[].contents | select(.) | keys | .[]' \
			"$filename" | sort -C || \
			err "$filename: $slice: \"contents\" entries are not sorted"
	done
}

for f in "$@"; do
	if [[ ! -f "$f" ]]; then
		err "file \"$f\" not found"
		continue
	fi
	check_slices_sorted "$f"
	check_essential_sorted "$f"
	check_contents_sorted "$f"
done

exit $EXIT_CODE
```

</details>

It works well, no problems there. The main problem is the speed of execution. The bash script is quite slow. I timed the bash script and the python script in this PR on the slices from the ubuntu-22.04 release. It had a total of 106 slice definition files. Here's the comparison:

1. Bash script

```
$ time ./lint.sh slices/*
warning: slices/base-files.yaml: slice names are not sorted
error: slices/base-files.yaml: base: "essential" entries are not sorted
error: slices/base-files.yaml: release-info: "contents" entries are not sorted
error: slices/base-passwd.yaml: data: "contents" entries are not sorted
warning: slices/bash.yaml: slice names are not sorted
error: slices/bash.yaml: bins: "contents" entries are not sorted
warning: slices/coreutils.yaml: slice names are not sorted
error: slices/dotnet-hostfxr-6.0.yaml: libs: "essential" entries are not sorted
warning: slices/fonts-dejavu-core.yaml: slice names are not sorted
error: slices/libasan6.yaml: libs: "essential" entries are not sorted
error: slices/libc6.yaml: libs: "contents" entries are not sorted
error: slices/libcurl4.yaml: libs: "essential" entries are not sorted
error: slices/libfontconfig1.yaml: libs: "essential" entries are not sorted
error: slices/libfreetype6.yaml: libs: "essential" entries are not sorted
error: slices/libglib2.0-0.yaml: libs: "essential" entries are not sorted
error: slices/libldap-2.5-0.yaml: libs: "contents" entries are not sorted
error: slices/libmount1.yaml: libs: "essential" entries are not sorted
error: slices/libnss3.yaml: libs: "contents" entries are not sorted
warning: slices/libpython3.11-stdlib.yaml: slice names are not sorted
error: slices/libpython3.11-stdlib.yaml: development-tools: "contents" entries are not sorted
error: slices/libpython3.11-stdlib.yaml: pydoc: "contents" entries are not sorted
error: slices/libunwind-13.yaml: libs: "contents" entries are not sorted
error: slices/libunwind8.yaml: libs: "contents" entries are not sorted
warning: slices/openjdk-8-jre-headless.yaml: slice names are not sorted
error: slices/openjdk-8-jre-headless.yaml: security: "essential" entries are not sorted
error: slices/openjdk-8-jre-headless.yaml: awt: "essential" entries are not sorted
error: slices/openjdk-8-jre-headless.yaml: core: "contents" entries are not sorted
error: slices/openjdk-8-jre-headless.yaml: security: "contents" entries are not sorted
error: slices/openjdk-8-jre-headless.yaml: awt: "contents" entries are not sorted
error: slices/openjdk-8-jre-headless.yaml: management: "contents" entries are not sorted
error: slices/openjdk-8-jre-headless.yaml: hprof: "contents" entries are not sorted
error: slices/openjdk-8-jre-headless.yaml: debug: "contents" entries are not sorted
error: slices/openjdk-8-jre-headless.yaml: tools: "contents" entries are not sorted
error: slices/openssl.yaml: bins: "essential" entries are not sorted
error: slices/python3.11.yaml: core: "essential" entries are not sorted
warning: slices/tzdata.yaml: slice names are not sorted
error: slices/tzdata.yaml: base: "contents" entries are not sorted
error: slices/tzdata.yaml: africa: "contents" entries are not sorted
error: slices/tzdata.yaml: america: "contents" entries are not sorted
error: slices/tzdata.yaml: config: "contents" entries are not sorted
error: slices/tzdata.yaml: etc: "contents" entries are not sorted
error: slices/tzdata.yaml: eurasia: "contents" entries are not sorted
error: slices/tzdata.yaml: pacific: "contents" entries are not sorted

real	1m9.005s
user	0m26.962s
sys	0m26.491s
```

2. Python script

```
$ time ./lint.py slices/*
slices/base-files.yaml
  error   base: "essential" entries are not sorted (--sorted-essential)
  error   release-info: "contents" entries are not sorted (--sorted-contents)
  error   slice names are not sorted (--sorted-slices)

slices/base-passwd.yaml
  error   data: "contents" entries are not sorted (--sorted-contents)

slices/bash.yaml
  error   bins: "contents" entries are not sorted (--sorted-contents)
  error   slice names are not sorted (--sorted-slices)

slices/coreutils.yaml
  error   slice names are not sorted (--sorted-slices)

slices/dotnet-hostfxr-6.0.yaml
  error   libs: "essential" entries are not sorted (--sorted-essential)

slices/fonts-dejavu-core.yaml
  error   slice names are not sorted (--sorted-slices)

slices/libasan6.yaml
  error   libs: "essential" entries are not sorted (--sorted-essential)

slices/libc6.yaml
  error   libs: "contents" entries are not sorted (--sorted-contents)

slices/libcurl4.yaml
  error   libs: "essential" entries are not sorted (--sorted-essential)

slices/libfontconfig1.yaml
  error   libs: "essential" entries are not sorted (--sorted-essential)

slices/libfreetype6.yaml
  error   libs: "essential" entries are not sorted (--sorted-essential)

slices/libglib2.0-0.yaml
  error   libs: "essential" entries are not sorted (--sorted-essential)

slices/libldap-2.5-0.yaml
  error   libs: "contents" entries are not sorted (--sorted-contents)

slices/libmount1.yaml
  error   libs: "essential" entries are not sorted (--sorted-essential)

slices/libnss3.yaml
  error   libs: "contents" entries are not sorted (--sorted-contents)

slices/libpython3.11-stdlib.yaml
  error   development-tools: "contents" entries are not sorted (--sorted-contents)
  error   pydoc: "contents" entries are not sorted (--sorted-contents)
  error   slice names are not sorted (--sorted-slices)

slices/libunwind-13.yaml
  error   libs: "contents" entries are not sorted (--sorted-contents)

slices/libunwind8.yaml
  error   libs: "contents" entries are not sorted (--sorted-contents)

slices/openjdk-8-jre-headless.yaml
  error   awt: "contents" entries are not sorted (--sorted-contents)
  error   awt: "essential" entries are not sorted (--sorted-essential)
  error   core: "contents" entries are not sorted (--sorted-contents)
  error   debug: "contents" entries are not sorted (--sorted-contents)
  error   hprof: "contents" entries are not sorted (--sorted-contents)
  error   management: "contents" entries are not sorted (--sorted-contents)
  error   security: "contents" entries are not sorted (--sorted-contents)
  error   security: "essential" entries are not sorted (--sorted-essential)
  error   slice names are not sorted (--sorted-slices)
  error   tools: "contents" entries are not sorted (--sorted-contents)

slices/openssl.yaml
  error   bins: "essential" entries are not sorted (--sorted-essential)

slices/python3.11.yaml
  error   core: "essential" entries are not sorted (--sorted-essential)

slices/tzdata.yaml
  error   africa: "contents" entries are not sorted (--sorted-contents)
  error   america: "contents" entries are not sorted (--sorted-contents)
  error   base: "contents" entries are not sorted (--sorted-contents)
  error   config: "contents" entries are not sorted (--sorted-contents)
  error   etc: "contents" entries are not sorted (--sorted-contents)
  error   eurasia: "contents" entries are not sorted (--sorted-contents)
  error   pacific: "contents" entries are not sorted (--sorted-contents)
  error   slice names are not sorted (--sorted-slices)


real	0m0.311s
user	0m0.310s
sys	0m0.000s
```

The bash script takes about a minute compared to 300 ms for the Python script. Of course, we are not out here to get the optimal speed. But do you think we should prefer the Python script since it's not very slow?

Reply:
Update: I was using the [snap](https://snapcraft.io/yq), which was very slow. Shifted to `go install` and `yq` is blazing fast again.

I will replace the python script with bash.

Reply:
Resolved in https://github.com/canonical/chisel-releases/pull/146/commits/894c8e389519213f2f6f580b660aad5f32fcdcd5.